### PR TITLE
Start Week 3 Project - Server Programming

### PR DESCRIPTION
**Pull request by ralucatudor**
Start week 3 project - **extend the portfolio** website built in week-2 to support user **comments**. 

Finished steps 1, 2 and 3 from the [walkthrough](https://github.com/ralucatudor/step/blob/master/walkthroughs/week-3-server/comments-walkthrough.md). Step 1 was a simple 1-line replacement. For **step 2** (Intro to Fetch), I created a new class (GreetingServlet) that returns a greeting - I added a button on the About Me section which calls, when clicked, a function that fetches the greeting from the server and adds it to the DOM.

For **step 3**, I added a section for **Comments in the navbar**. I created the **Comment class** which is instantiated in DataServlet.
For now, the comments are hard-coded (I followed the steps in the walkthrough). Basically step 3 is _"modify DataServlet.java to return some hard-coded JSON data, and then modify script.js to fetch that JSON and build a UI from it. "_

Live: [rtudor-step-2020.appspot.com/](http://rtudor-step-2020.appspot.com/)

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/GreetingServlet.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+/** Servlet that returns a greeting. */
+@WebServlet("/greeting")
+public class GreetingServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html;");
+    response.getWriter().println("Hello World!");
+  }
+}
```
Comment (skopecki) [inline portfolio/src/main/java/com/google/sps/servlets/GreetingServlet.java]: Nit: Add new line at the end of the document.

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -78,3 +85,25 @@ function getGallery() {
...
+ * Fetches comments from the server and adds them to the DOM.
+ */
+function getComments() {
+    const wrapper = document.getElementById("wrapper");
+    wrapper.textContent = '';
+    fetch('/data').then(response => response.json()).then((comments) => {
+        for (let comment of comments) {
+            wrapper.appendChild(
+                createListElement(comment.text + ', by ' + comment.author +
+                                  ', posted on: ' + comment.date));
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: The `<li>` tags should be used inside a list, so you would need an `<ol>` or `<ul>` tag which holds the list elements.

**Review by skopecki — APPROVED** [review]

File: portfolio/pom.xml [hunk 3]
```diff
@@ -23,6 +23,11 @@
       <version>4.0.1</version>
       <scope>provided</scope>
     </dependency>
+    <dependency>
```
Comment (ChangchengXiao) [inline portfolio/pom.xml]: Doesn't necessarily apply here. But it's normally a good idea to keep things sorted, e.g. in alphabetical order.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import java.util.Date;
+
+/** Class containing comment data. */
+public final class Comment {
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:20]: This is a classic Java value class. Thus it's better to use @AutoValue here.
https://mincong.io/2018/08/21/why-you-should-use-auto-value-in-java/

Note this is beyond the scope of your project. But it's good to learn about it.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,19 +14,38 @@
...
+/** Servlet that handles comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+
+    List<Comment> comments = new ArrayList<>();
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:34]: Normally we prefer to use immutable list in this case where we don't expect others to change the entry after the list is created.

Please take the following instructions for details:
https://docs.oracle.com/javase/9/core/creating-immutable-lists-sets-and-maps.htm#JSCOR-GUID-DD066F67-9C9B-444E-A3CB-820503735951

File: portfolio/src/main/java/com/google/sps/servlets/GreetingServlet.java (hunk 1 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/GreetingServlet.java]: My bad! Done, thanks!

File: portfolio/pom.xml (hunk 3 above)
Comment (ralucatudor) [reply portfolio/pom.xml:26]: I'll keep that in mind from now on, thank you!

Comment (ralucatudor) [reply portfolio/pom.xml]: I sorted the dependencies & I'll keep that tip in mind from now on, thank you!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -14,19 +14,38 @@
...
+
+    // Add a few hard-coded comments
+    comments.add(new Comment("Lorem ipsum dolor sit amet.", "Anonymus1"));
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:37]: If you use ImmutableList, they could be created by chained "ImmutableList.of".

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -14,19 +14,38 @@
...
+    // TODO: remove initialization in the final version
+
+    /** 
+     * Convert comments List into a JSON string using Gson library and
+     * send the JSON as the response
+     */
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(comments));
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:49]: Is this just for debugging?

Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:49]: No, here I'm sending the JSON as the response, so it's not for debug 😃

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 4 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:20]: Great tip, thanks a lot! Will do.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:34]: I didn't know I can do that in Java, so thank you for the info! Will do.

**Pull request merged**